stl/inc/xloctime

```diff
@@ -651,6 +651,17 @@ protected:
     __CLR_OR_THIS_CALL ~time_get_byname() noexcept override {}
 };
 
+constexpr bool _Is_valid_fmt_specifier(const char _Specifier) {
+    constexpr char _Valid_specifiers[] = "aAbBcCdDeFgGhHIjmMnprRStTuUVwWxXyYzZ";
+    for (const char _Valid_specifier : _Valid_specifiers) {

```

Comment:
This array will be recreated each call.
It's gonna be quite expensive at runtime, better make it `static`.

Reply:
I thought that making it `constexpr` would have the same effect? If not I will change it as part of the next commit when more reviews come in case I have to change something else.

Reply:
`constexpr` and `static` are quite orthogonal.
[Here](https://stackoverflow.com/a/13867690/4582609) is a good explanation.

Reply:
> This array will be recreated each call. It's gonna be quite expensive at runtime, better make it `static`.

But if we do so, the enclosing `_Is_valid_fmt_specifier` function itself can't be `constexpr` until C++23, see WG21-P2242R3 and WG21-P2647R1 (both of which are not yet implemented by MSVC at this moment).

Reply:
> But if we do so, the enclosing `_Is_valid_fmt_specifier` function itself can't be `constexpr`

Of course.
But this function is intended for runtime only, because the `put` method itself isn't `constexpr`. Or am I missing something?

Reply:
I haven't checked the codegen (I am a lazy kitty :cat2:) but my understanding of the compiler front-end and back-end tells me that it will emit code to push each element of the array on the stack, as I have observed previously - i.e. @jovibor is correct. This is because using an array makes it "address-taken". The compiler back-end can coalesce constants, but it has to preserve the observable guarantee that "distinct objects are at distinct addresses", so as soon as something is address-taken, the back-end will usually emit it as a distinct object.

Additionally, my understanding is that `constexpr` doesn't really increase the potential for optimizations in MSVC (other compilers may differ). MSVC doesn't consider it a hint that computations can be eagerly boiled down. The optimizer is able to do constant propagation, inlining, etc. and boil things down to constants (as long as the source code is available header-only, or LTCG gets iinvolved), but `constexpr` doesn't make that more likely.

Disclaimer: I'm not a compiler dev and could be wrong.

In practice, what this means for library code is that we try to consistently say `static constexpr` for arrays, but plain `constexpr` for scalar constants. With the noted exception (thanks @frederick-vs-ja) that if the function has to be `constexpr`, then that currently limits us from saying `static constexpr` (so annoying).

What we should *actually* do here is lift out the array to namespace scope, where it can be `_INLINE_VAR constexpr`. As I agreed in the other comment thread, optimizing the performance isn't critical here, but we may as well avoid gratuitous costs when doing so is cheap in terms of source code impact, and lifting out the array is very easy. This way, the valid characters will be "burned into ROM" in a single place that all template instantiations can use (`.rdata`, I think?), without the compiler emitting annoying instructions to push each character onto the stack before looping.

I'll validate and push changes.